src/protocolsupport/protocol/packet/middleimpl/clientbound/play/v_4_5_6_7_8_9r1_9r2_10_11_12r1_12r2_13/InventorySetItems.java

```diff
@@ -20,18 +20,34 @@ public InventorySetItems(ConnectionImpl connection) {
 	@Override
 	public RecyclableCollection<ClientBoundPacketData> toData() {
 		ProtocolVersion version = connection.getVersion();
-		if (version.isBefore(ProtocolVersion.MINECRAFT_1_9) && ((cache.getWindowCache().getOpenedWindow() == WindowType.PLAYER) || (windowId == 0))) {
-			itemstacks.remove(itemstacks.size() - 1);
-		} else if (version.isBefore(ProtocolVersion.MINECRAFT_1_9) && (cache.getWindowCache().getOpenedWindow() == WindowType.BREWING)) {
-			itemstacks.remove(4);
-		} else if (version.isBefore(ProtocolVersion.MINECRAFT_1_8) && (cache.getWindowCache().getOpenedWindow() == WindowType.ENCHANT)) {
-			itemstacks.remove(1);
+		switch (WindowSlotsRemappingHelper.getOpenedWindowType(connection, windowId)) {
+			case PLAYER: {
+				if (!WindowSlotsRemappingHelper.hasPlayerOffhandSlot(version)) {
+					itemstacks.remove(WindowSlotsRemappingHelper.PLAYER_OFF_HAND_SLOT);
+				}
+				break;
+			}
+			case BREWING: {
+				if (!WindowSlotsRemappingHelper.hasBrewingBlazePowderSlot(version)) {
+					itemstacks.remove(WindowSlotsRemappingHelper.BREWING_BLAZE_POWDER_SLOT);
+				}
+				break;
+			}
+			case ENCHANT: {
+				if (!WindowSlotsRemappingHelper.hasEnchantLapisSlot(version)) {
+					itemstacks.remove(WindowSlotsRemappingHelper.ENCHANT_LAPIS_SLOT);
+				}
+				break;
+			}
+			default: {
+				break;
+			}
```

Comment:
This really improves the readability of the code. Nice!

Reply:
yea, lots of good changes from shev 👍 got a bit more than a month out of date with upstream, wanted to get caught up. especially with the bigger changes